ma_gym/envs/traffic_junction/traffic_junction.py

```diff
@@ -125,25 +270,175 @@ def __create_grid(self):
 
         return _grid
 
+    def step(self, agents_action):
+        """
+        Performs an action in the environment and steps forward. At each step a new agent enters the road by
+        one of the 4 gates according to a probability "_arrive_prob". A "ncoll" reward is given to an agent if it
+        collides and all of them receive "-0.01*step_n" to avoid traffic jams.
+
+        :param agents_action: list of actions of all the agents to perform in the environment
+        :type agents_action: list
+
+        :return: agents observations, rewards, if agents are done and additional info
+        :rtype: tuple
+        """
+        assert len(agents_action) == self.n_agents, \
+            "Invalid action! It was expected to be list of {}" \
+            " dimension but was found to be of {}".format(self.n_agents, len(agents_action))
+
+        # Todo: Add check to validate each sub-action. It should be a valid action.
+        # This must be done before position of any car is updated
+
+        assert all([action_i in ACTION_MEANING.keys() for action_i in agents_action]), \
+            "Invalid action found in the list of sampled actions {}" \
+            ". Valid actions are {}".format(agents_action, ACTION_MEANING.keys())
+
+        self._step_count += 1
+        rewards = [0 for _ in range(self.n_agents)]  # initialize rewards array
+
+        # checks if there is a collision; this is done in the __update_agent_pos method
+        for agent_i, action in enumerate(agents_action):
+            if not (self._agent_dones[agent_i]) and self._on_the_road[agent_i]:
+                collision_flag = self.__update_agent_pos(agent_i, action)
+                if collision_flag:
+                    rewards[agent_i] += self._collision_reward
+
+
+        # adds new car according to the probability _arrive_prob
+        if random.uniform(0, 1) < self._arrive_prob:
+            free_gates = self.__is_gate_free()
+            # if there are agents outside the road and if any gate is free
+            if not all(self._on_the_road) and free_gates:
+                # then gets first agent on the list which is not on the road
+                agent_to_enter = self._on_the_road.index(False)
+                pos = random.choice(free_gates)
+                self._agents_direction[agent_to_enter] = self._route_vectors[pos]
+                self.agent_pos[agent_to_enter] = pos
+                self.curr_cars_count += 1
+                self._on_the_road[agent_to_enter] = True
+                self._agent_turned[agent_to_enter] = False
+                self._agents_routes[agent_to_enter] = random.randint(1, 3)
+                self.__update_agent_view(agent_to_enter)
+
+
+        # verifies if any destination place was reached
+        for agent_i in range(self.n_agents):
```

Comment:
Can this for-loop be merged with the for-loop on line #300 ?

Reply:
I think we can merge them but an agent that arrives to a gate will not receive the step penalty for the first arriving time step